<a href="https://colab.research.google.com/github/JayShekhavat/Diabetes-Prediction-ML-Project/blob/main/Diabetes_ML_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Import all libraries**

In [ ]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree  import DecisionTreeClassifier

#**Load dataset in csv formate**

In [ ]:
df = pd.read_csv("/content/diabetes (1).csv")

# **Show top 5 records**

In [ ]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

#**Show 5 sample records**

In [ ]:
df.sample(5)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
23             9      119             80             35        0  29.0   
322            0      124             70             20        0  27.4   
202            0      108             68             20        0  27.3   
374            2      122             52             43      158  36.2   
398            3       82             70              0        0  21.1   

     DiabetesPedigreeFunction  Age  Outcome  
23                      0.263   29        1  
322                     0.254   36        1  
202                     0.787   32        0  
374                     0.816   28        0  
398                     0.389   25        0

In [ ]:
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [ ]:
from scipy.stats import pointbiserialr

correlation1, p_value1 = pointbiserialr(df['Pregnancies'], df['Outcome'])
print(f"Correlation1: {correlation1}, P-Value1: {p_value1}")

correlation2, p_value2 = pointbiserialr(df['Glucose'], df['Outcome'])
print(f"Correlation2: {correlation2}, P-Value2: {p_value2}")

correlation3, p_value3 = pointbiserialr(df['BloodPressure'], df['Outcome'])
print(f"Correlation3: {correlation3}, P-Value3: {p_value3}")

correlation4, p_value4 = pointbiserialr(df['SkinThickness'], df['Outcome'])
print(f"Correlation4: {correlation4}, P-Value4: {p_value4}")

correlation5, p_value5 = pointbiserialr(df['Insulin'], df['Outcome'])
print(f"Correlation5: {correlation5}, P-Value5: {p_value5}")

correlation6, p_value6 = pointbiserialr(df['BMI'], df['Outcome'])
print(f"Correlation6: {correlation6}, P-Value6: {p_value6}")

correlation7, p_value7 = pointbiserialr(df['DiabetesPedigreeFunction'], df['Outcome'])
print(f"Correlation7: {correlation7}, P-Value7: {p_value7}")

correlation8, p_value8 = pointbiserialr(df['Age'], df['Outcome'])
print(f"Correlation8: {correlation8}, P-Value8: {p_value8}")


Correlation1: 0.22189815303398686, P-Value1: 5.065127298053635e-10
Correlation2: 0.466581398306874, P-Value2: 8.935431645289576e-43
Correlation3: 0.06506835955033283, P-Value3: 0.07151390009776264
Correlation4: 0.0747522319183194, P-Value4: 0.038347704820490915
Correlation5: 0.13054795488404775, P-Value5: 0.0002861864603603164
Correlation6: 0.29269466264444544, P-Value6: 1.2298074873116917e-16
Correlation7: 0.17384406565296007, P-Value7: 1.2546070101487771e-06
Correlation8: 0.2383559830271977, P-Value8: 2.209975460665451e-11


#**Show columns in dataset**

In [ ]:
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

#**Show shape of dataset**

In [ ]:
df.shape

(768, 9)

#**Show all statistics values of each columns of dataset**

In [ ]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

#**Show informarion of dataset**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


#**Check null values**

In [ ]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

###There is no any null values in dataset

#**Check duplicate values in dataset**

In [ ]:
df.duplicated().sum()

0

###There is no any duplicate values in dataset

In [ ]:
df.sample(5)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
652            5      123             74             40       77  34.1   
442            4      117             64             27      120  33.2   
233            4      122             68              0        0  35.0   
694            2       90             60              0        0  23.5   
685            2      129             74             26      205  33.2   

     DiabetesPedigreeFunction  Age  Outcome  
652                     0.269   28        0  
442                     0.230   24        0  
233                     0.394   29        0  
694                     0.191   25        0  
685                     0.591   25        0

#**Split dataset into train test**

In [ ]:
x = df.iloc[:,0:-1]
y = df.iloc[:,-1]

In [ ]:
x.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33

In [ ]:
y.head()

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=40)

In [ ]:
y_train.head()

223    0
203    0
112    0
507    0
577    1
Name: Outcome, dtype: int64

In [ ]:
x_test.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
370            3      173             82             48      465  38.4   
388            5      144             82             26      285  32.0   
611            3      174             58             22      194  32.9   
550            1      116             70             28        0  27.4   
232            1       79             80             25       37  25.4   

     DiabetesPedigreeFunction  Age  
370                     2.137   25  
388                     0.452   58  
611                     0.593   36  
550                     0.204   21  
232                     0.583   22

In [ ]:
y_test.head()

370    1
388    1
611    1
550    0
232    0
Name: Outcome, dtype: int64

In [ ]:
scaler = StandardScaler()
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.fit_transform(x_test)



##**K-nearest neighbor Algorithm**

In [ ]:
tree_model = KNeighborsClassifier(n_neighbors=11, p=2, metric='euclidean')
ml_tree_model = tree_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_tree_model.predict(x_test_scaled)

In [ ]:
tree_accuracy = accuracy_score(y_predicted, y_test)


In [ ]:
tree_accuracy

0.7077922077922078

###K-nearest neighbors algorithm give 0.74777... accuracy

##**Logistic regression algorithm**

In [ ]:
lr_model = LogisticRegression()
ml_lr_model = lr_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_lr_model.predict(x_test_scaled)
lr_accuracy = accuracy_score(y_predicted, y_test)
lr_accuracy

0.7662337662337663

###Logistic regression algorithm give 0.7663.. accuracy

##**Support Vector Machine algorithm**

In [ ]:
svc_model = SVC()
ml_svc_model = svc_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_svc_model.predict(x_test_scaled)
svc_accuracy = accuracy_score(y_predicted, y_test)
svc_accuracy

0.7402597402597403

###Support vector machine give 0.7410.. accuracy

##**Decision Trees algorithm**

In [ ]:
tree_model = DecisionTreeClassifier()
ml_tree_model = tree_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_tree_model.predict(x_test_scaled)
tree_accuracy = accuracy_score(y_predicted, y_test)
tree_accuracy

0.7467532467532467

###Decision tree classifier algorithm give 0.7468 accuracy

##**Conclusion**

###I used 4 algorithms, out of these algorithms Logistic regression algorithm give good accuracy than others algorithms.